nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/AbstractAMQPProcessor.java

```diff
@@ -0,0 +1,204 @@
+    public static final PropertyDescriptor V_HOST = new PropertyDescriptor.Builder()
+            .name("Virtual Host")
+            .description("Virtual Host name which segregates AMQP system for enhanced security.")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor USER = new PropertyDescriptor.Builder()
+            .name("User Name")
+            .description("User Name used for authentication and authorization.")
+            .required(true)
+            .defaultValue("guest")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor PASSWORD = new PropertyDescriptor.Builder()
+            .name("Password")
+            .description("Password used for authentication and authorization.")
+            .required(true)
+            .defaultValue("guest")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .sensitive(true)
+            .build();
+    public static final PropertyDescriptor AMQP_VERSION = new PropertyDescriptor.Builder()
+            .name("AMQP Version")
+            .description("AMQP Version. Currently only supports AMQP v0.9.1.")
+            .required(true)
+            .allowableValues("0.9.1")
+            .defaultValue("0.9.1")
+            .build();
+
+    static List<PropertyDescriptor> descriptors = new ArrayList<>();
+
+    /*
+     * Will ensure that list of PropertyDescriptors is build only once, since
+     * all other lifecycle methods are invoked multiple times
+     */
+    static {
+        Field[] fields = AbstractAMQPProcessor.class.getDeclaredFields();
+        try {
+            for (Field field : fields) {
+                field.setAccessible(true);
+                if (PropertyDescriptor.class.isAssignableFrom(field.getType())) {
+                    descriptors.add((PropertyDescriptor) field.get(null));
+                }
+            }
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to build PropertyDescriptor list", e);
+        }
+    }
+
+    protected volatile Connection amqpConnection;
+
+    protected volatile T targetResource;
+
+    /**
+     * Will builds target resource ({@link AMQPPublisher} or
+     * {@link AMQPConsumer}) upon first invocation and will delegate to the
+     * implementation of
+     * {@link #rendezvousWithAmqp(ProcessContext, ProcessSession)} method for
+     * further processing.
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        synchronized (this) {
+            this.buildTargetResource(context);
+        }
+        this.rendezvousWithAmqp(context, session);
+    }
+
+    /**
+     * Will close current AMQP connection.
+     */
+    @OnStopped
+    public void close() {
+        try {
+            this.targetResource.close();
```

Comment:
If the call to createConnection() were to throw an Exception, both targetResource and amqpConnection could be closed. Should check if these objects are null before calling close()